lib/improver/spotdata/extract_data.py

```diff
@@ -0,0 +1,374 @@
+
+        orography : numpy.array
+            Array of orography data on a grid that corresponds to the grid of
+            the the diagnostic cube.
+
+        Returns:
+        --------
+        Data extracted from the screen level temperature cube at spotdata sites
+        which has then been adjusted using a temperature lapse rate calculated
+        using local variations in temperature with orography.
+
+        """
+        def local_lapse_rate(cube, orography, node_list):
+            """
+            Least-squares fit to local temperature and altitude data for grid
+            points defined by node_list to calculate a local lapse rate.
+
+            """
+            y_data = cube.data[node_list]
+            x_data = orography[node_list]
+            matrix = np.vstack([x_data, np.ones(len(x_data))]).T
+            gradient, intercept = lstsq(matrix, y_data)[0]
+            return [gradient, intercept]
+
+        data = np.empty(shape=(len(sites)))
+
+        for i_site, site in enumerate(sites.itervalues()):
+            altitude = site['altitude']
+
+            i, j = neighbours['i'][i_site], neighbours['j'][i_site]
+            edgepoint = neighbours['edgepoint'][i_site]
+            node_list = nearest_n_neighbours(i, j, no_neighbours)
+            if edgepoint:
+                node_list = node_edge_test(node_list, cube)
+
+            llr = local_lapse_rate(cube, orography, node_list)
+            data[i_site] = llr[0]*altitude + llr[1]
+
+        return self.make_cube(cube, data, sites)
+
+    def model_level_temperature_lapse_rate(self, cube, sites, neighbours,
+                                           pressure_on_height_levels,
+                                           surface_pressure,
+                                           temperature_on_height_levels):
+        """
+        Lapse rate method based on potential temperature. Follows the work of
+        S.B. Vosper 2005 (Near-surface temperature variations over complex
+        terrain).
+
+        Args:
+        -----
+        cube : iris.cube.Cube
+            A cube of screen level temperatures at a single time.
+
+        sites/neighbours : See process() above.
+
+        pressure_on_height_levels : iris.cube.Cube
+            Pressure data at 50m above the surface at an equivalent time to the
+            cube of screen level temperatures.
+
+        surface_pressure : iris.cube.Cube
+            Cube of surface pressures at an equivalent time to the cube of
+            screen level temperatures.
+
+        temperature_on_height_levels : iris.cube.Cube
+            Temperature data at 50m above the surface at an equivalent time to
+            the cube of screen level temperatures.
+
+        Returns:
+        --------
+        Data extracted from the screen level temperature cube at spotdata sites
+        which has then been adjusted using a temperature lapse rate calculated
+        using multi-level data.
+
```

Comment:
What a fantastic docstring! You're right I'm not sure how to deal with this level of information within the code, which is something that we'll have to revisit, but thanks for putting the effort in! 🌟